wcfsetup/install/files/lib/system/worker/UserContentRemoveWorker.class.php

```diff
@@ -128,19 +146,29 @@ private function generateData() {
 		
 		foreach ($contentProviders as $contentProvider) {
 			if ($this->contentProvider === null || (is_array($this->contentProvider) && in_array($contentProvider->objectType, $this->contentProvider))) {
-				/** @var IUserContentProvider $processor */
-				$processor = $contentProvider->getProcessor();
-				$contentList = $processor->getContentListForUser($this->user);
-				$count = $contentList->countObjects();
-				
-				if ($count) {
-					$this->data['provider'][$contentProvider->objectType] = [
-						'count' => $count,
-						'objectTypeID' => $contentProvider->objectTypeID,
-						'nicevalue' => $contentProvider->nicevalue ?: 0
-					];
+				foreach ($this->user as $user) {
+					/** @var IUserContentProvider $processor */
+					$processor = $contentProvider->getProcessor();
+					$contentList = $processor->getContentListForUser($user);
+					$count = $contentList->countObjects();
 					
-					$this->data['count'] += ceil($count / $this->limit) * $this->limit;
+					if ($count) {
+						if (!isset($this->data['provider'][$contentProvider->objectType])) {
+							$this->data['provider'][$contentProvider->objectType] = [
+								'count' => 0,
+								'objectTypeID' => $contentProvider->objectTypeID,
+								'nicevalue' => $contentProvider->nicevalue ?: 0,
+								'user' => [],
+							];
+						}
+						
+						$this->data['provider'][$contentProvider->objectType]['user'][$user->userID] = [
+							'count' => $count,
+						];
+						$this->data['provider'][$contentProvider->objectType]['count'] += $count;
```

Comment:
This is a horrible data structure and I believe this is part of the reason the issue happened. May I suggest the following:

1. Sort the providers by niceValue before iterating them.
     - They are already in the correct order then.
2. Invert the 2 loops.
     - This processes all contents for a single user before processing the next user, feeling more natural.
3. Then make this a flat queue:
     ```
     $this->queue[] = [ 'userID' => …, 'provider' => …, 'remaining' => … ];
     ```

When processing you then simply take the first item in the queue and process it until `remaining < 0`, then removing it and proceeding with the next. No need for complex nesting. The number of remaining items then is a simple `array_sum(array_column($this->queue, 'remaining'))` away.

Reply:
I think it complexifies the logic way more, then it is now. Especially because we currently only execute the object types for the users that really have content (and object types w/o any content will be skipped anyway). In a flat map, I need to test that with every object type again. Keep in mind that the logic has been in use for 2 (?) versions.

Reply:
> Especially because we currently only execute the object types for the users that really have content (and object types w/o any content will be skipped anyway).

You are free to not put the entries into the queue if `count == 0`.

> Keep in mind that the logic has been in use for 2 (?) versions.

The previous logic has been in use for several versions, but the logic to handle multiple users is new.

And quite frankly: I'm having a very hard time following the control flow here due to `$this->data` containing 5 (!) nested arrays where you need to subtract the number of processed items in various locations to make sure to correctly remove specific parts of the arrays to indicate that this part was fully processed.

A queue is as simple of a data structure as you can get. It _will_ drastically simplify the code: Process the first `(userID, objectTypeID)` pair until `count` is non-positive, repeat until no more pairs exist.